lua_modules/liquidcrystal/NTest_lc_i2c4bit.lua

```diff
@@ -0,0 +1,116 @@
+-- Run LiquidCrystal through some basic tests
+--
+--[[   The simplest method of testing leaves everything in RAM and is just
+
+  TCLLIBPATH=./tests/expectnmcu expect ./tests/tap-driver.expect \
+    ./lua_tests/mispec.lua \
+    ./lua_modules/liquidcrystal/liquidcrystal.lua \
+    ./lua_modules/liquidcrystal/lc-i2c4bit.lua \
+    ./lua_modules/liquidcrystal/NTest_lc_i2c4bit.lua
+
+Better to put NTest and LiquidCrystal in LFS, and then just run
+
+  TCLLIBPATH=./tests/expectnmcu expect ./tests/tap-driver.expect \
+    ./lua_modules/liquidcrystal/NTest_lc_i2c4bit.lua
+
+--]]
+
+
+local N = ...
+N = (N or require "NTest")("liquidcrystal-i2c4bit")
+
+local metalcd
+local metaback
+local backend
+local lcd
+
+collectgarbage()
+print("HEAP init", node.heap())
+
+N.test("import constructor", function()
+  metalcd = require "liquidcrystal"
+  collectgarbage() print("HEAP constructor imported ", node.heap())
+  ok(metalcd, "constructor")
+end)
+
+N.test("import backend", function()
+  metaback = require "lc-i2c4bit"
+  collectgarbage() print("HEAP backend imported ", node.heap())
+  ok(metaback, "backend")
+end)
+
+N.test("construct backend", function()
+  backend = metaback({
+   address = 0x27,
+   id  = 0,
+   speed = i2c.SLOW,
+   sda = 2,
+   scl = 1,
+  })
+  collectgarbage() print("HEAP backend built", node.heap())
+end)
+
+N.test("construct lcd", function()
+  lcd = metalcd(backend, false, true, 20)
+  collectgarbage() print("HEAP lcd built", node.heap())
+end)
+
+N.test("release backend", function()
+  backend = nil
+  collectgarbage() print("HEAP backend released", node.heap())
+end)
+
+N.test("Wait for LCD OK", function() while lcd:busy() do end end)
+
+N.test("custom character", function()
+  local glyph = { 0x1F, 0x15, 0x1B, 0x15, 0x1F, 0x10, 0x10, 0x0 }
+  lcd:customChar(0, glyph)
+  ok(eq(glyph,lcd:readCustom(0)), "read back custom character")
+end)
+
+N.test("draw", function()
+  lcd:cursorMove(0)
+  lcd:write("abc")
+  lcd:cursorMove(10,1)
+  lcd:write("de")
+  lcd:cursorMove(10,2)
+  lcd:write("fg")
+  lcd:cursorMove(12,3)
+  lcd:write("hi\000")
+  lcd:cursorMove(18,4)
+  lcd:write("jk")
+end)
+
+N.test("read back 'a'", function() lcd:home()           ok(eq(0x61, lcd:read())) end)
```

Comment:
Tests should be self contained and not rely on the tests that run before.
If you want to print something for a test try using code like
`ok(eq(0x61, lcd:read()), "read back 'a'")`

Reply:
Er, well, full independence is going to be tricky, but I reworked things.